### PR TITLE
Add a new `augment-vis` saved object type

### LINKED ISSUE #2893: [Feature Anywhere] Add a `augment-vis` saved object type
As part of #2880, we propose to create a new saved object type called `augment-vis`. The purpose of this new type will be to store the relationships between a plugin resource (e.g., anomaly detector or alerting monitor) to a visualization. The main components/fields include:
1. Plugin resource ID (e.g., detector id)
2. A reference to a `vis_layer` expression function, including the name of the function and the args needed (for more details on this expression function, see #2894)
3. A reference to the visualization that can be stored in the `references` field.

To create and register this new type, we can do the following:
Client/public-side:
- [ ] Create a saved object class that extends `SavedObjectClass`
- [ ] Create a saved object loader that extends `SavedObjectLoader` to use as a wrapper for performing CRUD operations on the new saved object class / type
- [ ] Create `extractReferences`/`injectReferences` helper fns for creating and parsing the stored `references` to the saved objects
- [ ] Instantiate the loader in the `start` lifecycle of dependent plugins (e.g., visualizations plugin, downstream plugins like Anomaly Detection), such that they can use the loader for creating/fetching the saved objects
- [ ] Register in the `saved_objects_management` plugin (this may not be needed)

Server-side:
- [ ] Create a saved objects type that extends `SavedObjectsType`, used for defining how the object as managed, and how the particular mappings will be stored in the system index
- [ ] Register this type in the `setup` lifecycle of the plugin

Open questions:
1. ~~Should this new type be registered & integrated with the `saved_objects_management` plugin? This will allow visibility into what is stored in the saved objects, but ideally users don't need or want access to it. Making changes to it may result in weird behavior~~ UPDATE: yes we will support it see [here](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3109#discussion_r1055837516)
2. ~~Should this new type be created in a standalone plugin, or in an existing plugin? If in an existing plugin, such as `Visualizations`, it allows for less boilerplate code, but may not be logical if the `feature-anywhere` type is used for integration with other saved objects in the future, such as dashboards, saved searches, index patterns, other new saved object types, etc.~~ UPDATE: it will live in a standalone plugin - see #2958

### PR DESCRIPTION
**Pull request by ohltyler**
### Description
This PR adds a new `augment-vis` saved object type. Details on the design can be found in the [related issue](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2893)

Specifically, this PR makes the following changes:

Client-side:
- creates & defines an augment vis saved object class
- creates helper fns to convert from client-side definition (with no references) to server-side definition (with references) & vice-versa
- introduce a saved object loader for the new saved object type. This loader provides a clean way to interact with all saved objects of this type, and can be consumed in dependent plugins for performing CRUD operations on saved objects of this type
- creates and sets the new loader in the `vis_augmenter` plugin's `start` lifecycle step

Server-side:
- creates & defines a augment vis saved object type + index mapping
- registers the new type in the `vis_augmenter` plugin's `setup` lifecycle step

It also registers the saved object in the Saved Objects Management plugin, by:
- adding a management section in the `augment-vis` saved object server-side definition,
- registering a capabilities provider to allow some of the actions to work in the management plugin, and
- adding the type to the plugin registry

Screenshots of an example `augment-vis` type in the Saved Objects Management plugin:

Fig. 1: new object is highlighted (note there is no link to the title since there is no dedicated plugin for this new saved object):
Screen Shot 2023-01-04 at 2 02 25 PM

Fig. 2: viewing the relationship of the saved object, showing the visualization saved object as the child:
Screen Shot 2023-01-04 at 2 02 52 PM

### Issues Resolved
Closes #2893 
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by ohltyler** [comment]
Keeping as draft until #3108 has been reviewed & merged; I will then rebase this PR

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:56]: @joshuarrrr @ashwin-pc any opinions on making this saved object type visible in the saved objects management plugin or not? 

I'm leaning towards the latter, due to the fact that it's a bit more "internal" than existing saved object types, and isn't meant to be edited by users. Also it may scale to the point where there is a large amount of them that could clutter the UI.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:24]: nit - I'd recommend extracting this string as a const

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:29]: Is there a reason you want an `object` for this? Other saved objects I've seen just use `text` and store a JSON serialized blob. But having a well defined object may actually be better, depending on how you plan to use it.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:42]: I'm not sure you want/need an empty string here? In general, when the id is null or undefined, it means the object is new, and the defaults will be assigned.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:57]: agree, no need to provide an edit path

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:55]: I'm not sure you want/need this either, if we're not enabling an edit path.

**Review by joshuarrrr — COMMENTED** [review]
Not a comprehensive review yet,  but this is looking good so far. Just a few minor comments and questions.

File: src/plugins/visualize/public/plugin.ts
Comment (ohltyler) [reply src/plugins/visualize/public/plugin.ts:201]: I'm actually going to remove any dependencies with the `visualize` plugin for now. Current scope is to not support this feature in the edit vis view but this is an ongoing discussion.

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:16]: `index: false` means the field is stored but not indexed or available for search. `doc_values: false` means the field does not support aggregation or sorting, or accessible via a script, which saves disk space. Because `visName` is an internal value set via `injectReferences()`/`extractReferences()`, it is only needed to be stored. This follows the pattern of `savedSearchRefName` in the `visualization` saved object: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/visualizations/server/saved_objects/visualization.ts#L63

File: src/plugins/vis_augmenter/public/types.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/types.ts:14]: Right, it's the client-side interface which maps the `visName` stored in the mappings to the `visId` in the corresponding reference via `injectReferences()`. It follows the same pattern as the `visualization` saved object interface - see https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/visualizations/public/types.ts#L68-L69

File: src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts:6]: removed

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:19]: Good catch, yes. Added in latest commit

File: src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts:16]: `SavedObjectLoader.get()` requires the arg to be of type `Record<string, unknown>`. Would you suggest to change this fn arg to be of the same type? The reason I preferred to keep the arg as `ISavedAugmentVis` was to make it cleaner on the plugin-developer side which would be calling this function, and knowing what the expected input would be.

File: src/plugins/vis_augmenter/public/index.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/index.ts:14]: updated

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:24]: updated

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:29]: Good point. I think I will switch to `text` and store a JSON serialized blob. My concern with having a well-defined object for this is complications around fluidity on the `args` field, and I don't think there's any reason to make it more typesafe here.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:42]: I was following the pattern of visualization saved objects here:
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/visualizations/public/saved_visualizations/_saved_vis.ts#L115

But I see it is different for dashboard saved objects: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/dashboard/public/saved_dashboards/saved_dashboard.ts#L100-L101

Is there one you would prefer over the other? I don't have a strong opinion since both auto-produce ids when it is a new obj getting created.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts]: Leaving this open until final determination on showing in saved objects management plugin is decided.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:58]: Leaving this open until final determination on showing in saved objects management plugin is decided.

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (ashwin-pc) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts]: Yeah if we have a scaling issue, we can find UI solutions to mitigate that, but that page is an advanced feature, so limiting the users ability to manage saved objects by themselves feels limiting.

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts]: Got it, thanks for the inputs. With that I agree it makes sense to expose it here; I can work on that as a follow-up in this PR.

**Comment by ohltyler** [comment]
@ashwin-pc @joshuarrrr the [latest commit](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3109/commits/97382772544aa76e789d8ecebbf4ef048d979b01) registers the new type in the management plugin. The plugin framework provides all of the functionality out-of-the-box but I've also manually confirmed the CRUD operations work as expected. The main changes needed were adding a `management` section in the server-side definition, registering a capabilities provider to allow some of the actions to work in the management plugin, and adding the type to the plugin registry.

I also added a helper `formatTitle` fn to clean up formatting of the toast when there is no `title` attribute on the saved object, since I noticed this issue when testing out my changes. Before it would show `undefined`

File: src/plugins/saved_objects_management/public/lib/in_app_url.ts
Comment (kristenTian) [inline src/plugins/saved_objects_management/public/lib/in_app_url.ts:42]: just curious, are we using this `show` for augment-vis? I recall it seems to be a registered capability.

File: src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis.ts
Comment (kristenTian) [inline src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis.ts:39]: why remove these attributes before return?

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis.ts:39]: I was thinking these would become unnecessary fields once the vis ID has been extracted, similar to what's done in [injectReferences()](https://github.com/ohltyler/OpenSearch-Dashboards/blob/add-saved-obj/src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts#L53).

But given the fn signature is only about adding fields to the source obj, and for consistency with other implementations of `mapHitSource()`, I agree that these shouldn't be removed. I'll update

File: src/plugins/saved_objects_management/public/lib/in_app_url.ts
Comment (ohltyler) [reply src/plugins/saved_objects_management/public/lib/in_app_url.ts:42]: It's required for the inspection/edit to work. Otherwise there is an NPE since it tries to access capabilities for this new type, and nothing is set.

File: src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis.ts]: Fixed in [this commit](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3109/commits/f1b08adfe2aebed6cab9755980e5f909877f0aa4)

**Comment by ashwin-pc** [comment]
@ohltyler pretty neat implementation, can you also as a part of this PR document the process to create a new saved object type?

@abbyhu2000 Can you take a look at this PR too since you have a lot of context about saved objects?

**Comment by ohltyler** [comment]
> @ohltyler pretty neat implementation, can you also as a part of this PR document the process to create a new saved object type? (optional but highly appreciated!)

Sure, I'll go ahead and add a section in the README 😄

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:44]: nah, I don't know or remember enough to have a strong opinion either way. The explicit pattern you have is probably more obvious, so better from that respect.

**Comment by ohltyler** [comment]
@ashwin-pc @kristenTian @joshuarrrr can I get another look at this PR? I'd like to get this one merged next to help unblock the plugin teams.

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:15]: so these commented lines are just to show the defaults?

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:18]: This will be the same for any augmentations linked to the same vis, right? That doesn't seem ideal. Do we have a plugin resource name we could also include for disambiguation? In either case, this isn't a blocker, but something we may want to open an issue and come back to later.

Comment (joshuarrrr) [inline src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:18]: nit - prefer template literals over string concatenation

```suggestion
      return `Augment-${obj.attributes.visName}`;
```

File: src/plugins/vis_augmenter/public/saved_augment_vis/utils/test_helpers.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/utils/test_helpers.ts:25]: out of curiosity - was this a pattern you picked up from elsewhere, or something new you decided would be useful?

File: src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts:15]: Yeah, I like the usage of `ISavedAugmentVis`. Just seems like it should be simpler for the typescript compiler to understand that it's compatible with `Record<string, unknown>`.

File: src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts:36]: Is `_0` meant to be auto-incrementing? Doesn't seem like something we want to hard-code, but I may be missing something.

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:15]: I'll remove these lines. Basically we can utilize the saved objects management plugin for the defaults on the icon and the search field, no need to specify/override here.

File: src/plugins/vis_augmenter/public/saved_augment_vis/utils/test_helpers.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/utils/test_helpers.ts:25]: I saw roughly similar patterns for existing saved object loaders - example: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/saved_objects/public/saved_object/saved_object_loader.test.ts#L23-L28

File: src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts:36]: It's the standard for how saved objects inject/extract references. It could be auto-incrementing, in the case of [saved dashboards](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/dashboard/public/saved_dashboards/saved_dashboard_references.ts#L51) regarding panel references, or just hardcoded as 0 if there is a 1:1 mapping between a saved obj and a reference, like in [visualization saved objects](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/visualizations/public/saved_visualizations/saved_visualization_references.ts#L64) when saving a saved search.

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:18]: yeah, I like that better!

File: src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/server/saved_objects/augment_vis.ts:15]: cool - it's fine to leave them, I was just making sure I understood.

File: src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts:36]: that all makes sense - can we add a comment to that effect?

**Review by joshuarrrr — APPROVED** [review]

**Review by kristenTian — APPROVED** [review]

**Comment by joshuarrrr** [comment]
merging despite build and verify failure on windows. See https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3400

**Pull request merged**
